File: apps/game/scripts/export-models.mjs
// Regenerate the authored GLBs in apps/game/public/models/ ({items,building_kit,
// props}.glb) from apps/game/assets/items.blend via headless Blender, then
// compress them with the version-pinned gltf-transform CLI (meshopt + WebP).
//
//   pnpm --filter @worldspring/game models:export
//
// A MANUAL authoring step (run after editing items.blend), NOT a CI step — CI has
// no Blender. Point at a specific Blender with BLENDER=/path/to/blender; otherwise
// the macOS app bundle and a `blender` on PATH are tried. See export-models.py for
// the per-collection / at-origin rules and docs/plans/research/ for the asset
// conventions CharacterRig.ts consumes (node name == ItemType, +Y up).
import { spawnSync } from "node:child_process";
import { existsSync, statSync } from "node:fs";
import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";

const scriptsDir = dirname(fileURLToPath(import.meta.url));
const gameRoot = join(scriptsDir, ".."); // apps/game
const blendFile = join(gameRoot, "assets", "items.blend");
const outDir = join(gameRoot, "public", "models");
const pyScript = join(scriptsDir, "export-models.py");

const candidates = [
  process.env.BLENDER,
  "/Applications/Blender.app/Contents/MacOS/Blender",
  "blender", // on PATH (Linux/CI-less authoring boxes)
].filter((c) => typeof c === "string" && c.length > 0);

// "blender" is accepted as a PATH lookup; absolute candidates must exist.
const blender = candidates.find((c) => c === "blender" || existsSync(c));
if (!blender) {
  console.error(
    `models:export: Blender not found. Install it, or set BLENDER=/path/to/blender.\n  tried: ${candidates.join(", ")}`,
  );
  process.exit(1);
}
if (!existsSync(blendFile)) {
  console.error(`models:export: source blend missing at ${blendFile}`);
  process.exit(1);
}

console.log(`models:export: ${blender}\n             ${blendFile} -> ${outDir}`);
const r = spawnSync(
  blender,
  ["--background", blendFile, "--python", pyScript, "--", outDir],
  { stdio: "inherit" },
);
if (r.error) {
  console.error(`models:export: failed to launch Blender — ${r.error.message}`);
  process.exit(1);
}
if ((r.status ?? 1) !== 0) process.exit(r.status ?? 1);

// Post-export: gltf-transform optimize (meshopt + 512px WebP), version-pinned
// via the @gltf-transform/cli devDependency — same settings as the Meshy asset
// pipeline (crate.glb, items/*.glb). Structural passes stay OFF: the runtime
// fetches nodes by name and clones them (CharacterRig.ts conventions), so
// flatten/join/instance/palette must not rewrite the scene graph, and simplify
// is pointless on these low-poly meshes. drei's useGLTF decodes meshopt by
// default, proven in-app by the already-compressed crate.glb.
const gltfTransform = join(gameRoot, "node_modules", ".bin", "gltf-transform");
// Keep in sync with MANIFEST in export-models.py.
const EXPORTED_GLBS = ["items.glb", "building_kit.glb", "props.glb"];

if (!existsSync(gltfTransform)) {
  console.error(
    "models:export: gltf-transform CLI missing — run `pnpm install` (@gltf-transform/cli is a devDependency of @worldspring/game)",
  );
  process.exit(1);
}
for (const fname of EXPORTED_GLBS) {
  const glbPath = join(outDir, fname);
  // Blender exiting 0 does not guarantee every expected GLB landed (e.g. this
  // list drifting from the .py MANIFEST) — fail with a clear message instead
  // of an ENOENT stack trace from statSync.
  if (!existsSync(glbPath)) {
    console.error(
      `models:export: expected ${fname} missing from ${outDir} after Blender export — check MANIFEST in export-models.py`,
    );
    process.exit(1);
  }
  const before = statSync(glbPath).size;
  const o = spawnSync(
    gltfTransform,
    [
      "optimize",
      glbPath,
      glbPath,
      "--compress", "meshopt",
      "--texture-compress", "webp",
      "--texture-size", "512",
      "--flatten", "false",
      "--join", "false",
      "--instance", "false",
      "--palette", "false",
      "--simplify", "false",
    ],
    { stdio: ["ignore", "ignore", "inherit"] },
  );
  if (o.error || (o.status ?? 1) !== 0) {
    console.error(
      `models:export: gltf-transform optimize failed for ${fname}${o.error ? ` — ${o.error.message}` : ""}`,
    );
    process.exit(o.status ?? 1);
  }
  const after = statSync(glbPath).size;
  console.log(
    `models:export: optimized ${fname} ${(before / 1024).toFixed(1)}KB -> ${(after / 1024).toFixed(1)}KB`,
  );
}
process.exit(0);

File: apps/game/scripts/export-models.py
# Headless Blender GLB exporter — run via scripts/export-models.mjs
# (`pnpm --filter @worldspring/game models:export`), NOT directly.
#
# One collection per GLB (apps/game/assets/items.blend holds Items / BuildingKit
# / Props; the `Collection` of camera+lights is ignored). For each, every
# TOP-LEVEL object's location is zeroed before export: the blend lays objects out
# in a grid for authoring, but the runtime fetches each node by name and clones it
# AT ORIGIN — held items positioned by GRIP_TRANSFORMS (CharacterRig.ts), world
# props by placement. Parented children (e.g. the deer's legs, parented to `deer`)
# keep their relative offset, so assemblies stay intact. The .blend on disk is
# never modified (headless, no save). Khronos glTF, +Y up, no Draco/meshopt — the
# conventions CharacterRig.ts documents (node name == ItemType, business-end +Z).
import bpy
import os
import sys

# Args after the `--` separator: [out_dir]
_argv = sys.argv[sys.argv.index("--") + 1 :] if "--" in sys.argv else []
OUT_DIR = _argv[0] if _argv else os.path.join("apps", "game", "public", "models")

# Authored collection -> emitted GLB filename.
MANIFEST = {
    "Items": "items.glb",
    "BuildingKit": "building_kit.glb",
    "Props": "props.glb",
}


def export_collection(name: str, out_path: str) -> bool:
    coll = bpy.data.collections.get(name)
    if coll is None:
        print(f"[export-models] ERROR: no collection '{name}' in the blend")
        return False
    # Zero each TOP-LEVEL object's location for the export, then RESTORE it in a
    # finally — so this stays pure: it never leaves the in-memory grid layout
    # mutated, even across the other collections' exports or when run inside an
    # interactive Blender (only `location` is touched; rotation/scale are not).
    saved = {obj: obj.location.copy() for obj in coll.all_objects if obj.parent is None}
    try:
        for obj in saved:
            obj.location = (0.0, 0.0, 0.0)
        bpy.ops.object.select_all(action="DESELECT")
        for obj in coll.all_objects:
            obj.select_set(True)
        bpy.ops.export_scene.gltf(
            filepath=out_path,
            export_format="GLB",
            use_selection=True,
            export_yup=True,
        )
    finally:
        for obj, loc in saved.items():
            obj.location = loc
    print(f"[export-models] {name} -> {out_path} ({len(coll.all_objects)} objects)")
    return True


os.makedirs(OUT_DIR, exist_ok=True)
ok = True
for coll_name, fname in MANIFEST.items():
    if not export_collection(coll_name, os.path.join(OUT_DIR, fname)):
        ok = False
sys.exit(0 if ok else 1)
